### PR TITLE
Automatically preload index files that are both tiny and very hot.

### PR DESCRIPTION
**Pull request by jpountz**
The default codec has a number of small and hot files, that actually used to be fully loaded in memory before we moved them off-heap. In the general case, these files are expected to fully fit into the page cache for things to work well. Should we give control over preloading to codecs? This is what this commit does for the following files:
 - Terms index (`tip`)
 - Points index (`kdi`)
 - Stored fields index (`fdx`)
 - Terms vector index (`tvx`)

This only has an effect on `MMapDirectory`.

### CODE REVIEW

**Review by uschindler — APPROVED** [review]
In general I am not against this. Looks technically correct.

But I am not sure if tis really helps for all cases. We load the index files on startup and therefore accept an extra delay that could be used for something else. On the other hand the files can still be purged by the OS kernel, because "preload" is a one-time action.

For MemeorySegmentIndexInput I have a solution, because I was already waiting for some change like this (I had a similar idea to add a new IOContent). With Project Panama and when you pass some special command line option to allow unsafe libc access, we can get a MethodHandle that allows us to lock the MemorySegment into page cache, so it can't be purged. This in combination with preload (where I am not sure if thats a good idea!) could be a good idea. If you are interested I had some slides

So in short: The addition of the new IOContext is fine, but I tend to now implement Panama native access to mlock2 the memory segment! See a future PR by me.

See this code of Java 17 times (looks a bit different in 19/20), I had this in a talk 1.5 years ago:
image

**Comment by rmuir** [comment]
I think preload is different from mlock, mlock needs way more discussion and personally I'm against it. mlock would be an operational hassle because of default resource limits on linux as well.

File: lucene/core/src/java/org/apache/lucene/store/MMapDirectory.java [hunk 1]
```diff
@@ -235,7 +235,7 @@ public IndexInput openInput(String name, IOContext context) throws IOException {
     ensureOpen();
     ensureCanRead(name);
     Path path = directory.resolve(name);
-    return PROVIDER.openInput(path, context, chunkSizePower, preload, useUnmapHack);
+    return PROVIDER.openInput(path, context, chunkSizePower, preload || context.load, useUnmapHack);
```
Comment (uschindler) [inline lucene/core/src/java/org/apache/lucene/store/MMapDirectory.java:238]: See my last message: It may be better in future to leave the decission to preload or map to the implementation of the provider. For exactly that reason the context is passed to the provider!

So I would move the `|| context.load` to somewhere in MappedByteBufferIndexInputProvider and MemorySegmentIndexInputProvider so they can have different code.

Maybe also rename "load" to something else because it is up to implementation how to guarantee how something is "sticked to memory".

**Comment by uschindler** [comment]
> I think preload is different from mlock, mlock needs way more discussion and personally I'm against it. mlock would be an operational hassle because of default resource limits on linux as well.

This would of course be optional (it must be optional as it also requires extra JVM command line params). I think we should give users the option to do it.

Personally I also tend to be against "load".

File: lucene/core/src/java/org/apache/lucene/store/MMapDirectory.java (hunk 1 above)
Comment (jpountz) [reply lucene/core/src/java/org/apache/lucene/store/MMapDirectory.java:238]: > So I would move the || context.load to somewhere in MappedByteBufferIndexInputProvider and MemorySegmentIndexInputProvider so they can have different code.

Do you mean making it a separate argument?

Comment (uschindler) [reply lucene/core/src/java/org/apache/lucene/store/MMapDirectory.java:238]: I would have moved the decission to those 2 places, no extra param needed:
- https://github.com/apache/lucene/blob/57ac311c709da2f398ea73896eae158b1126c5b4/lucene/core/src/java/org/apache/lucene/store/MappedByteBufferIndexInputProvider.java#L77
- https://github.com/apache/lucene/blob/57ac311c709da2f398ea73896eae158b1126c5b4/lucene/core/src/java19/org/apache/lucene/store/MemorySegmentIndexInputProvider.java#L54

But thats not important. This issue here was exactly "planned" when I made the provider's signature. Currently the context is not used downstream.

Comment (uschindler) [reply lucene/core/src/java/org/apache/lucene/store/MMapDirectory.java:238]: So for now keep it as is!

**Comment by jpountz** [comment]
@uschindler I think I understand Robert's concerns against mlock, but less yours against loading. Do you think that it would be useless to preload if these files can then get paged out? My thinking was that most users would rather like Lucene to take a bit longer to open indices if it helps provide better performance on initial queries, noting that it wouldn't make opening much slower given how tiny these files tend to be compared to the overall dataset (~0.17% of the total index size for the index generated by nightly benchmarks). In general these files would stay in the page cache all the time, but if they don't because of unusual access patterns or competing processes and the page cache decides to page out some data, it wouldn't be much worse than today. I do think some users will see appeal in the mlock option and getting stronger guarantees that this data wouldn't get paged out, but this feels like something that should be an opt-in rather than the default?

**Comment by uschindler** [comment]
I think the predicate would also simplify the current Elasticserach code which instantiates 2 MMapDirectory instances with the same underlying dir and uses FileSwitchDir in front. A predicate is much better to handle this ("which files to preload").

**Comment by jpountz** [comment]
Agreed, this would be a welcome simplification.

**Comment by uschindler** [comment]
we can keep and deprecate the old boolean setter that delegates to 2 predefined "alaways on", "always off" predicates.

**Comment by jpountz** [comment]
@uschindler Something like that?

**Comment by uschindler** [comment]
Looks fine. I would add 3 predefined constants:

```java
public static final BiPredicate<String,Context> ALL_FILES = (file, context) -> true;
public static final BiPredicate<String,Context> NO_FILES = (file, context) -> false;
public static final BiPredicate<String,Context> USE_LOAD_CONTEXT = (file, context) -> context.load;
```

Then you could also keep the deprecated getter by compaing with ALL_FILES; The Default would be the new one.

What's missing is a little bit of better documentation in `setPreload()` and possibly the class javadocs. It is unknown what the first param "String" of the predicate is, it should document that it is the filename.

**Comment by uschindler** [comment]
Unrelated problem with Github CI:

> org.apache.lucene.index.TestIndexFileDeleter.testExcInDecRef

this test failed for me yesterday in the same way. Seems to be a bug.

**Comment by uschindler** [comment]
looks great. I don't like the `PRELOAD_` prefix on the constants, because I plan to use the same for mlock2. I was already thinking about this, but came to no good names. So my proposal was the above.

**Comment by jpountz** [comment]
OK. I wanted to use names that give a sense of when these general `BiPredicate` objects are useful without looking up javadocs, but your point about reusing for mlock makes sense.

File: lucene/core/src/java/org/apache/lucene/store/MMapDirectory.java [hunk 2]
```diff
@@ -85,25 +85,23 @@ public class MMapDirectory extends FSDirectory {
...
    * Argument for {@link #setPreload(BiPredicate)} that configures all files that use the {@link
    * IOContext#LOAD} to be preloaded upon opening them. This is the default.
    */
-  public static final BiPredicate<String, IOContext> PRELOAD_BASED_ON_IO_CONTEXT =
+  public static final BiPredicate<String, IOContext> BASED_ON_IO_CONTEXT =
       (filename, context) -> context.load;
 
   private boolean useUnmapHack = UNMAP_SUPPORTED;
-  private BiPredicate<String, IOContext> preload = PRELOAD_BASED_ON_IO_CONTEXT;
+  private BiPredicate<String, IOContext> preload = BASED_ON_IO_CONTEXT;
```
Comment (uschindler) [inline lucene/core/src/java/org/apache/lucene/store/MMapDirectory.java:104]: maybe `BASED_ON_LOAD_IO_CONTEXT`

**Comment by uschindler** [comment]
Looks good to me, i think maybe say for the constant that it uses LOAD context.

**Comment by uschindler** [comment]
...namining is hard...

**Comment by uschindler** [comment]
I will open an issue in Solr to adapt their code.

**Comment by rmuir** [comment]
> @uschindler I think I understand Robert's concerns against mlock, but less yours against loading. Do you think that it would be useless to preload if these files can then get paged out? My thinking was that most users would rather like Lucene to take a bit longer to open indices if it helps provide better performance on initial queries, noting that it wouldn't make opening much slower given how tiny these files tend to be compared to the overall dataset (~0.17% of the total index size for the index generated by nightly benchmarks). In general these files would stay in the page cache all the time, but if they don't because of unusual access patterns or competing processes and the page cache decides to page out some data, it wouldn't be much worse than today. I do think some users will see appeal in the mlock option and getting stronger guarantees that this data wouldn't get paged out, but this feels like something that should be an opt-in rather than the default?

I'm not against preloading i just don't think it does anything here. It just madvises the kernel for read-ahead and then touches every page.

Seems something went badly wrong here: if these files need to be hot, then just load the structures into the heap? all this talk about preloading and locking pages is not good. why were these "small hot files" removed from the heap?

**Review by uschindler — APPROVED** [review]

**Comment by jpountz** [comment]
I reverted the bits that would load some files in physical memory by default. These files don't have to be in memory and I actually like that these files are not in the heap because it makes lots of things easier, e.g. you don't have to size our heap depending on how much data you're dealing with - 0.17% is still 1.75GB of memory if you have 1TB worth of index data.

**Comment by rmuir** [comment]
And 1.75GB is the same amount if you want to mlock the pages. Which no operating system is gonna let you do by default, on linux it will require modifying system soft/hard limits.

If a user has 1TB worth of index data, is it more intuitive that they should run with e.g. at least a 4GB heap? Or that they should go mucking around in /etc/security?

**Comment by rmuir** [comment]
also wanna note my concern around any use of mlock, that, users may not know how to do the right thing and be lazy and just run as root. this is sadly the stuff that users do, i'd hate to encourage it.

so to me it seems like the heap is more user-friendly way to get the performance that you want.

File: lucene/core/src/java/org/apache/lucene/store/MMapDirectory.java [hunk 3]
```diff
@@ -93,15 +93,8 @@ public class MMapDirectory extends FSDirectory {
    */
   public static final BiPredicate<String, IOContext> NO_FILES = (filename, context) -> false;
 
-  /**
-   * Argument for {@link #setPreload(BiPredicate)} that configures all files that use the {@link
-   * IOContext#LOAD} to be preloaded upon opening them. This is the default.
-   */
-  public static final BiPredicate<String, IOContext> BASED_ON_LOAD_IO_CONTEXT =
```
Comment (uschindler) [inline lucene/core/src/java/org/apache/lucene/store/MMapDirectory.java]: We can still provide this option.

**Comment by uschindler** [comment]
BTW, sorry that I mentioned mlock2... oh man oh man.

File: lucene/core/src/java/org/apache/lucene/store/IOContext.java [hunk 4]
```diff
@@ -45,8 +45,8 @@ public enum Context {
   public final boolean readOnce;
 
   /**
-   * This flag indicates that the file is expected to be heavily accessed in a random-access fashion
-   * and should have its content in memory.
+   * This flag indicates that the file is expected to be heavily accessed in a random-access
```
Comment (uschindler) [inline lucene/core/src/java/org/apache/lucene/store/IOContext.java:48]: the name of the field is not really applicable for that description. Maybe it should be called different? I dont know a good name, e.g. "tinyRandomAccess"

Comment (uschindler) [reply lucene/core/src/java/org/apache/lucene/store/IOContext.java:48]: But what is tiny? 1.6 Gigabytes?!?!

**Comment by uschindler** [comment]
Maybe we should start again here:
- sorry for mentioning mlock2, I will think of a way in the future to hook into the MemorySegment and configure it in native ways (e.g. set madvise) on MMapDirectory. I won't implement a methodhandle to mlock2 without haveing the fun of fighting with @rmuir ;-]
- Maybe split the PR into two parts: (1) adding the improved preload BiPredicate (this would allow Elasticsearch to do whatever they want) and (2) add the IOContext support for those files and also add the example as proposed here.

I would not change the default. That's up to the end-user of Lucene (e.g. Elasticsearch, Solr, Opensearch).

**Comment by rmuir** [comment]
Well it was good to mention mlock as that "makes more sense" than just "preload" (which has a confusing name IMO). "preload" just faults it in but if there is memory pressure it can then just be paged out, so that's why i don't think it "does anything".

**Comment by uschindler** [comment]
I think to get a "real sticky preload to heap", we would need to add a NRTCachingDirectory-like wrapper on top of any Directory that copies the file into ByteBuffersIndexInput.

**Comment by jpountz** [comment]
Agreed with your suggestion to reset and split into two PRs @uschindler. I'll close this one and open two new ones.

**Comment by jpountz** [comment]
This got split into #11929 and #11930.

**Comment by mikemccand** [comment]
> this test failed for me yesterday in the same way. Seems to be a bug.

This test has been failing for some time -- we have #10878 open for it.  I'll try to make some time to figure out what's happening.
